node/src/isolate/external.cpp

```diff
@@ -192,41 +192,93 @@ struct metrics_load_t {
     apply();
 };
 
+
+// All async methods should capture shared_from_this()
+class connection_t : public std::enable_shared_from_this<connection_t> {
+    context_t& context_;
+    std::unique_ptr<logging::logger_t> log_;
+
+    std::unique_ptr<tcp::socket> socket_;
+    asio::deadline_timer connect_timer_;
+
+public:
+    using async_completion = std::function<void(const std::error_code&, std::shared_ptr<session_t>)>;
+
+    connection_t(context_t& context, const std::string& name, asio::io_service& io)
+        : context_(context)
+        , log_(context.log("universal_isolate/connection/" + name))
+        , socket_(std::make_unique<tcp::socket>(io))
+        , connect_timer_(io)
+    {}
+
+    auto
+    async_connect(tcp::endpoint endpoint, boost::posix_time::milliseconds timeout, async_completion&& callback) -> void
+    {
+        COCAINE_LOG_INFO(
+            log_, "connecting to external isolation daemon at {}", boost::lexical_cast<std::string>(endpoint));
+
+        const auto self = shared_from_this();
+
+        connect_timer_.expires_from_now(timeout);
+        connect_timer_.async_wait([self](const std::error_code& ec) {
+            COCAINE_LOG_INFO(self->log_, "timeout connecting to isolate daemon, ec = {}", ec);
+
+            // Socket could be moved out in `async_connect` handler, if not yet,
+            // or on any timer error just close it.
+            if (self->socket_) {
+                self->socket_->close();
+            }
+        });
+
+        socket_->async_connect(std::move(endpoint), [self, callback=std::move(callback)](const std::error_code& ec) {
+            std::shared_ptr<session_t> session;
+
+            if (self->socket_->is_open() && !ec) {
+                session = self->context_.engine().attach(std::move(self->socket_), nullptr);
+            }
+
+            callback(ec, std::move(session));
+        });
+    }
+};
+
 struct external_t::inner_t :
     public std::enable_shared_from_this<inner_t>
 {
     context_t& context;
     asio::io_service& io_context;
-    std::unique_ptr<tcp::socket> socket;
-    asio::deadline_timer connect_timer;
     std::shared_ptr<session_t> session;
     const std::string name;
     dynamic_t args;
     std::unique_ptr<logging::logger_t> log;
     std::atomic<id_t> cur_id;
     std::shared_ptr<dispatch<io::context_tag>> signal_dispatch;
     bool prepared;
-    bool connecting;
 
     std::chrono::time_point<std::chrono::system_clock> last_failed_connect_time;
     std::chrono::milliseconds seal_time;
+
     static constexpr std::chrono::milliseconds min_seal_time {1000ul};
     static constexpr std::chrono::milliseconds max_seal_time {1000000ul};
 
     std::vector<std::shared_ptr<spool_load_t>> spool_queue;
     std::vector<std::shared_ptr<spawn_load_t>> spawn_queue;
 
+    std::atomic_flag connecting_ = ATOMIC_FLAG_INIT;
```

Comment:
`connecting_` is public member, do not need to use an underline at the end of name

Reply:
Yep, should fix.